File: Liftr/RouteLineStringDecimation.swift
import CoreLocation
import Foundation

enum RouteLineStringDecimation {
    static let maxStoredCoordinates = 2000

    static func decimate(_ coords: [CLLocationCoordinate2D]) -> [CLLocationCoordinate2D] {
        guard coords.count > maxStoredCoordinates else { return coords }
        let maxPoints = maxStoredCoordinates
        let n = coords.count
        var out: [CLLocationCoordinate2D] = []
        out.reserveCapacity(maxPoints)
        let denom = maxPoints - 1
        for i in 0..<maxPoints {
            let idx = (i * (n - 1)) / denom
            out.append(coords[idx])
        }
        return out
    }

    static func encodeGeoJSONLineString(_ coords: [CLLocationCoordinate2D]) -> String? {
        guard coords.count >= 2 else { return nil }
        let parts = coords.map { "[\($0.longitude),\($0.latitude)]" }
        return "{\"type\":\"LineString\",\"coordinates\":[\(parts.joined(separator: ","))]}"
    }
}

File: android/app/src/main/java/com/lilru/liftr/externalroute/RouteLineStringDecimation.kt
package com.lilru.liftr.externalroute

data class LatLng(val latitude: Double, val longitude: Double)

object RouteLineStringDecimation {
    const val MAX_STORED_COORDINATES = 2000

    fun decimate(coords: List<LatLng>): List<LatLng> {
        if (coords.size <= MAX_STORED_COORDINATES) return coords
        val maxPoints = MAX_STORED_COORDINATES
        val n = coords.size
        val denom = maxPoints - 1
        return buildList(maxPoints) {
            for (i in 0 until maxPoints) {
                val idx = (i * (n - 1)) / denom
                add(coords[idx])
            }
        }
    }

    fun encodeGeoJsonLineString(coords: List<LatLng>): String? {
        if (coords.size < 2) return null
        val parts = coords.joinToString(",") { "[${it.longitude},${it.latitude}]" }
        return """{"type":"LineString","coordinates":[$parts]}"""
    }
}
